conformance/service_import.go

```diff
@@ -214,3 +215,50 @@ func testClusterIPServiceImport() {
 			reportNonConformant(fmt.Sprintf("The value %q is not a valid IP", serviceImport.Spec.IPs[0])))
 	})
 }
+
+func testHeadlessServiceImport() {
+	t := newTestDriver()
+
+	BeforeEach(func() {
+		t.helloService.Spec.ClusterIP = corev1.ClusterIPNone
+	})
+
+	JustBeforeEach(func() {
+		t.helloService.Spec.ClusterIP = corev1.ClusterIPNone
+		t.createServiceExport(&clients[0])
+	})
+
+	Specify("Exporting a headless service should create a ServiceImport of type Headless in the service's namespace in each cluster. "+
+		"Unexporting should delete the ServiceImport", Label(RequiredLabel), func() {
+		AddReportEntry(SpecRefReportEntry, "https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#service-types")
+
+		for i := range clients {
+			serviceImport := t.awaitServiceImport(&clients[i], helloServiceName, nil)
+			Expect(serviceImport).NotTo(BeNil(), reportNonConformant(fmt.Sprintf("ServiceImport was not found on cluster %q",
+				clients[i].name)))
+
+			Expect(serviceImport.Spec.Type).To(Equal(v1alpha1.Headless), reportNonConformant(
+				fmt.Sprintf("ServiceImport on cluster %q has type %q", clients[i].name, serviceImport.Spec.Type)))
+		}
+
+		By("Unexporting the service")
+
+		t.deleteServiceExport(&clients[0])
+
+		for i := range clients {
+			t.awaitNoServiceImport(&clients[i], helloServiceName, fmt.Sprintf(
+				"the ServiceImport still exists on cluster %q after unexporting the service", clients[i].name))
+		}
+	})
+
+	Specify("No clusterset IP should be allocated for a Headless ServiceImport", Label(RequiredLabel), func() {
+		AddReportEntry(SpecRefReportEntry, "https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#clustersetip")
+
+		serviceImport := t.awaitServiceImport(&clients[0], t.helloService.Name, nil)
+		Expect(serviceImport).NotTo(BeNil(), "ServiceImport was not found")
+
+		Consistently(func() []string {
+			return t.getServiceImport(&clients[0], t.helloService.Name).Spec.IPs
+		}).Within(5*time.Second).ProbeEvery(time.Second).Should(BeEmpty(), reportNonConformant(""))
```

Comment:
A default service seems to set both `clusterIP` and `clusterIPs` to the `None` string so I think we should do the same in a headless service (also the existing controller does that IIUC.

Reply:
For a `Service`, `clusterIP` is set to `None` by the user to specify that it's headless. I'm not sure why they did it this way rather than a define a service type. However, `ServiceImport` defines the `Headless` type so setting the `ips` field to `[]{"None"}` isn't necessary (and would seem a bit awkward). Also the MCS spec does not say it should be set to `None`. In fact [this section](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#service-types) in the "_ClusterIP: none (Headless)_" bullet, it says "_Headless services ... do not configure a VIP_" which indicates it should be empty.

> (also the existing controller does that IIUC)

Which controller are you referring to? If it's the reference implementation in this repo, I'm really familiar with the code but it does not explicitly set the `ServiceImport` `ips` field to `None` anywhere. The [ServiceImportReconciler](https://github.com/kubernetes-sigs/mcs-api/blob/8690854b517f61128d8603af11f417692f46930c/pkg/controllers/serviceimport.go#L57) explicitly ignores headless services. OTOH, the [ServiceReconciler](https://github.com/kubernetes-sigs/mcs-api/blob/8690854b517f61128d8603af11f417692f46930c/pkg/controllers/service.go#L71), which sets the `ips` field, does not ignore headless services and blindly copies the `service.Spec.ClusterIP` field which would propagate the `None` value although I think that's an oversight.